tests/cmdline/commands/test_process.py

```diff
@@ -490,3 +393,84 @@ def test_multiple_processes(self):
         self.assertIn('No callers found', get_result_lines(result)[0])
         self.assertIn(str(self.node_root.pk), get_result_lines(result)[1])
         self.assertIn(str(self.node_root.pk), get_result_lines(result)[2])
+
+
+@pytest.mark.skip(reason='fails to complete randomly (see issue #4731)')
+@pytest.mark.requires_rmq
+@pytest.mark.usefixtures('with_daemon')
+@pytest.mark.usefixtures('clear_database_before_test')
+@pytest.mark.parametrize('cmd_try_all', (True, False))
+def test_pause_play_kill(cmd_try_all):
+    """
+    Test the pause/play/kill commands
+    """
+    # pylint: disable=no-member, too-many-locals
+    from aiida.cmdline.commands.cmd_process import process_pause, process_play, process_kill
+    from aiida.manage.manager import get_manager
+    from aiida.engine import ProcessState
+    from aiida.orm import load_node
+
+    runner = get_manager().create_runner(rmq_submit=True)
+    cli_runner = CliRunner()
```

Comment:
I made a specific fixture for running click commands. It is in `tests/cmdline/commands/conftest.py` and is called `run_cli_command`. Search in existing tests to see how it is used. It also makes it a lot easier to check the results. Would be good to use this fixture every time you call a command in this test

Reply:
Haha, funny thing. Originally when I moved the test to pytest I just removed all the daemon setups and was actually using `run_cli_command`; however when I rebased there were a couple of conflicts with the stuff I removed and since in the middle of this was the problem with the test failing randomly, I didn't want to risk it so [I put it all back](https://github.com/aiidateam/aiida-core/pull/4671/commits/c6fc336b96b4194b0fc329e3b2214a54bab25d5d) the way it was before. Then basically with the `with_daemon` fixture and now with this one I basically took all out again xD...

Anyways: ok, I'll change it, but what do you mean it will "make it a lot easier to check the results"? The way I did it before I just replaced all the `cli_runner.invoke` with `run_cli_command` and removed the setup you are pointing in here, but the asserts I kept all the same no?

Reply:
It provides some additional functionality that makes the checking more succinct, for example checking for an exception, you can replace
```
result = cli_runner.invoke(command, [non_existing_process_id])
import traceback
assert result.exception is None, ''.join(traceback.format_exception(*result.exc_info))
assert 'Error:' in result.output
```
with
```
result = run_cli_command(command, [non_existing_process_id], raises=True)
assert 'Error:' in result.output
```
Also there is an attribute on the `result` called `output_lines` which already explodes the output string on the newline characters which is often something one needs to do in checking. Maybe not in this test though

Reply:
We could also think of adding a `match` argument to it like `pytest.raises` to take a regex to match the exception message against

Reply:
Ok, from looking at the code of `run_cli_command`, more specifically this part:

```python
        if raises:
            assert result.exception is not None, result.output
            assert result.exit_code != 0
        else:
            assert result.exception is None, ''.join(traceback.format_exception(*result.exc_info))
            assert result.exit_code == 0, result.output
```

I am assuming I can also remove these lines:

```python
    result = run_cli_command(process_pause, ['--wait', str(calc.pk)])
    assert result.exception is None, result.output # <---- This can be removed
    assert calc.paused
```

However, there is something that feels backwards: the `assert result.exception is not None, result.output` was being used when there was no exception, but inside `run_cli_command` it is called when `raises = True` (and viceverse for line `assert result.exception is None, ''.join(traceback.format_exception(*result.exc_info))`). Am I misunderstanding something or should these be switched?

> We could also think of adding a match argument to it like pytest.raises to take a regex to match the exception message against

Do you mean like, now for this PR =P? This test just does a single check of the output (the one you quoted), so maybe we can leave this change for a change where it would be used more. I could open an issue with a low priority tag.

Reply:
> I am assuming I can also remove these lines:

Yes you can remove the explicit check that the exception is `None` because that is done automatically.

> However, there is something that feels backwards: the `assert result.exception is not None, result.output` was being used when there was no exception, but inside `run_cli_command` it is called when `raises = True` (and viceverse for line `assert result.exception is None, ''.join(traceback.format_exception(*result.exc_info))`). Am I misunderstanding something or should these be switched?

To me the logic in `run_cli_command` makes sense. If the caller says `raises=True` they expect the command to raise, therefore we check that the `exception is not None` and if that is not true (i.e. the exception _was_ `None` and so the command didn't actually fail although it should have) then we print the output so the caller can inspect it to debug. Conversely, if `raises=False`, we make sure that the command did not except or else we show the output.

> > We could also think of adding a match argument to it like pytest.raises to take a regex to match the exception message against
> 
> Do you mean like, now for this PR =P? This test just does a single check of the output (the one you quoted), so maybe we can leave this change for a change where it would be used more. I could open an issue with a low priority tag.

No, not necessary for this PR. Was just an idea for a future feature. Indeed low priority issue would be great.

Reply:
> To me the logic in `run_cli_command` makes sense. If the caller says `raises=True` they expect the command to raise, therefore we check that the `exception is not None` and if that is not true (i.e. the exception _was_ `None` and so the command didn't actually fail although it should have) then we print the output so the caller can inspect it to debug. Conversely, if `raises=False`, we make sure that the command did not except or else we show the output.

Mmm ok, but then the second code snippet on your second comment should not have the `raises = True`, correct? The first snippet seems to be expecting no raise. In the current develop code it is slightly different but also seems to be expecting there to be no raise:

```python
result = self.cli_runner.invoke(command, [non_existing_process_id])
self.assertClickResultNoException(result)
self.assertIn('Error:', result.output)

# Defined elsewhere
def assertClickResultNoException(self, cli_result):  # pylint: disable=invalid-name
        self.assertIsNone(cli_result.exception, ''.join(traceback.format_exception(*cli_result.exc_info)))
```

> No, not necessary for this PR. Was just an idea for a future feature. Indeed low priority issue would be great.

:ok_hand: